llvm/lib/Transforms/IPO/RepoPruning.cpp

```diff
@@ -123,14 +123,71 @@ repodefinition::DigestType toDigestType(pstore::index::digest D) {
   return Digest;
 }
 
+/// Two kinds of pruning exist in the repository level and the module level. The
+/// repository level pruning: if the global object is compiled before and exists
+/// in the repository, it will be pruned. The module level pruning: in the same
+/// compilation unit (module), some global objects have the same digests. The
+/// RepoPruning pass could prune them as well.
+///
+/// This function is used to collect the GO's pruning status which include three
+/// possible states: in the repository, in the ModuleFragments and not in both.
+/// If GO is in the repository, collect the GO's xfixup names. If it is not in
+/// repository and not in ModuleFragments, put it into the ModuleFragments.
+///
+/// \param Repository The database.
+/// \param FragmentIndex A database fragment index.
+/// \param ModuleFragments The collection of fragments that will appear in this
+///        compilation but not include the pruned entries.
+/// \param GO A global object.
+/// \returns a tuple containing 1) true if the GO is in the existing repository.
+///          2) true if the GO doesn't exist in the existing repository
+///          but is in the ModuleFragments, and 3) the GO's digest.
+static std::tuple<bool, bool, pstore::index::digest> hasExistingGO(
+    const pstore::database &Repository,
+    std::shared_ptr<const pstore::index::fragment_index> const &FragmentIndex,
+    std::map<pstore::index::digest, const GlobalObject *> &ModuleFragments,
+    const GlobalObject &GO) {
```

Comment:
I'm not sure about this function signature. The function’s name (hasExistingGO) suggests that it will return a boolean indicating whether there is an existing GO. However, it returns no less than four things: a 3-tuple and sometimes mutates the map given by the third argument.

For me, this is actually less clear than the original code.

Reply:
Ok, I see. I have split the function into two functions and moved some codes back to the initial code. Hope the changes make clearer than the original code.